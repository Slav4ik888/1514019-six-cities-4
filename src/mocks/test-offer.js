export const testOffer = {
  id: 1,
  isPremium: true,
  isFavourite: false,
  previewImage: `img/apartment-03.jpg`,
  pictures: [
    `img/apartment-01.jpg`,
    `img/apartment-02.jpg`,
    `img/apartment-03.jpg`,
    `img/room.jpg`,
    `img/studio-01.jpg`,
    `img/studio-photos.jpg`],
  amenities: [`Wifi`, `Heating`, `Kitchen`, `Cable TV`],
  bedrooms: 4,
  maxGuestsNumber: 5,
  description: `This cozy and complete apartment in the heart of New York is a typical, traditional and authentic feel for how New Yorkers live. I welcome all guests to stay at this place during their stay and visit of New York. 
  Please note that this apartment is on the sixth floor and there is no lift (elevator).
  The apartment is convenient for all guests, fitted with a kitchen, small dining area and two bedrooms – couples and groups of friends will find this accommodation most optimal.`,
  host: {
    photo: `img/avatar-angelina.jpg`,
    name: `Angelina Jonson`,
    super: true
  },
  price: 180,
  rating: 5,
  cardTitle: `Beautiful & luxurious apartment at great location`,
  offerType: `House`,
  coordinates: [48.856663, 2.351556],
  city: {
    zoom: 8,
    coordinates: [48.856663, 2.351556],
  },
  location: {
    zoom: 8,
  },
  reviews: [
    {
      id: 1,
      author: {
        photo: `img/avatar-angelina.jpg`,
        name: `Strange jober`
      },
      description: `small dining area and two bedrooms – couples and groups of friends will find this accommodation`,
      date: 1593095836000,
      rating: 3
    },
    {
      id: 2,
      author: {
        photo: ``,
        name: `Slava`
      },
      description: `dining area and five bedrooms – couples and groups of friends will find this accommodation`,
      date: 1593095936000,
      rating: 4.7
    },
  ],

};
