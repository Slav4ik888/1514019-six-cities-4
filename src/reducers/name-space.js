export const NameSpace = {
  DATA: `DATA`,
  TRAVEL: `TRAVEL`,
  USER: `USER`,
  FAVORITE: `FAVORITE`,
};
