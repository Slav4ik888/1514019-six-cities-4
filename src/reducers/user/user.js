const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  authInfo: {},
};

const ActionType = {
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  AUTH_INFO: `AUTH_INFO`,
};

const ActionCreator = {
  requireAuthorization: (status) => {
    return {
      type: ActionType.REQUIRED_AUTHORIZATION,
      payload: status,
    };
  },
  setActiveAuth: (data) => {
    return {
      type: ActionType.AUTH_INFO,
      payload: data,
    };
  }
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then((res) => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.setActiveAuth(res.data));
        // console.log('USER  get /login: ', res.data);
      })
      .catch((err) => {
        throw err;
      });
  },

  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.login,
      password: authData.password,
    })
      .then((res) => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.setActiveAuth(res.data));
        // console.log('USER  post /login: ', res.data);
      })
      .catch((err) => {
        // console.log('err: ', err);
        throw err;
      });
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRED_AUTHORIZATION:
      return Object.assign({}, state, {
        authorizationStatus: action.payload,
      });
    case ActionType.AUTH_INFO:
      return Object.assign({}, state, {
        authInfo: action.payload,
      });
  }

  return state;
};

export {
  ActionCreator,
  AuthorizationStatus,
  ActionType,
  Operation,
  reducer,
};
