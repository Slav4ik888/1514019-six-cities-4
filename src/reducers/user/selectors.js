import {NameSpace} from '../name-space.js';

const NAME_SPACE = NameSpace.USER;

export const getUserStatus = (state) => {
  return state[NAME_SPACE].authorizationStatus;
};

export const getAuthInfo = (state) => {
  return state[NAME_SPACE].authInfo;
};
