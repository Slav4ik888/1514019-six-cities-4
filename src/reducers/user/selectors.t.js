import {getUserStatus, getAuthInfo} from './selectors.js';

const mockState = {
  authorizationStatus: `AUTH`,
  authInfo: {login: 2},
};

describe(`User reducer USER selector tests`, () => {
  it(`getUserStatus selector return correct value`, () => {
    expect(getUserStatus(mockState)).toEqual(`AUTH`);
  });
  it(`getAuthInfo selector return correct value`, () => {
    expect(getAuthInfo(mockState)).toEqual({login: 2});
  });
});
