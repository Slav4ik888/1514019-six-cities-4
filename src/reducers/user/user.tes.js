import {reducer, Operation, ActionCreator, ActionType, AuthStatus} from './user.js';
import MockAdapter from 'axios-mock-adapter';
import {createAPI} from '../../api.js';

// const userInfo = {
//   avatarUrl: `/static/avatar/7.jpg`,
//   email: `korzan.va@mail.ru`,
//   id: 1,
//   isPro: false,
//   name: `Vyacheslav`,
// };

it(`USER Reducer without additional parameters should be return initiaState`, () => {
  expect(reducer(void 0, {})).toEqual({
    authorizationStatus: AuthStatus.NO_AUTH,
    authInfo: {},
  });
});

it(`Reducer should change authorizationStatus by a given value`, () => {
  expect(reducer({
    authorizationStatus: AuthStatus.NO_AUTH,
  }, {
    type: ActionType.REQUIRED_AUTHORIZATION,
    payload: AuthStatus.AUTH,
  })).toEqual({
    authorizationStatus: AuthStatus.AUTH,
  });

  expect(reducer({
    authorizationStatus: AuthStatus.AUTH,
  }, {
    type: ActionType.REQUIRED_AUTHORIZATION,
    payload: AuthStatus.NO_AUTH,
  })).toEqual({
    authorizationStatus: AuthStatus.NO_AUTH,
  });

  expect(reducer({
    authorizationStatus: AuthStatus.AUTH,
  }, {
    type: ActionType.REQUIRED_AUTHORIZATION,
    payload: AuthStatus.AUTH,
  })).toEqual({
    authorizationStatus: AuthStatus.AUTH,
  });

  expect(reducer({
    authorizationStatus: AuthStatus.NO_AUTH,
  }, {
    type: ActionType.REQUIRED_AUTHORIZATION,
    payload: AuthStatus.NO_AUTH,
  })).toEqual({
    authorizationStatus: AuthStatus.NO_AUTH,
  });
});

it(`Reducer should change authInfo by a given value`, () => {
  expect(reducer({
    authInfo: {},
  }, {
    type: ActionType.AUTH_INFO,
    payload: {
      email: `korzan.va@mail.ru`,
      password: `fghjdk`,
    },
  })).toEqual({
    authInfo: {
      email: `korzan.va@mail.ru`,
      password: `fghjdk`,
    },
  });

  expect(reducer({
    authInfo: {
      email: `korzan.va@mail.ru`,
      password: `fghjdk`,
    },
  }, {
    type: ActionType.AUTH_INFO,
    payload: {},
  })).toEqual({
    authInfo: {},
  });
});

describe(`Action creators work correctly`, () => {
  it(`Action creator for require authorization returns correct action`, () => {
    expect(ActionCreator.requireAuthorization(AuthStatus.NO_AUTH)).toEqual({
      type: ActionType.REQUIRED_AUTHORIZATION,
      payload: AuthStatus.NO_AUTH,
    });

    expect(ActionCreator.requireAuthorization(AuthStatus.AUTH)).toEqual({
      type: ActionType.REQUIRED_AUTHORIZATION,
      payload: AuthStatus.AUTH,
    });
  });

  it(`Action creator for setActiveAuth returns correct action`, () => {
    expect(ActionCreator.setActiveAuth({
      email: `korzan.va@mail.ru`,
      password: `fghjdk`,
    })).toEqual({
      type: ActionType.AUTH_INFO,
      payload: {
        email: `korzan.va@mail.ru`,
        password: `fghjdk`,
      },
    });

    expect(ActionCreator.setActiveAuth({})).toEqual({
      type: ActionType.AUTH_INFO,
      payload: {},
    });
  });
});

const api = createAPI(() => {});

describe(`USER Operation work correctly`, () => {
  it(`Operation checkAuth returns correct action`, () => {
    const apiMock = new MockAdapter(api);
    const dispatch = jest.fn();
    const checkAuth = Operation.checkAuth();

    apiMock
      .onGet(`/login`) // Чтобы мок на запрос `/login`
      .reply(200, {}); // вернул ответ 200 и объект {}

    return checkAuth(dispatch, () => {}, api)
      .then(() => {
        expect(dispatch).toHaveBeenCalledTimes(2);
        expect(dispatch.mock.calls[0][0]).toEqual({
          type: ActionType.REQUIRED_AUTHORIZATION,
          payload: `AUTH`});
      });
  });

  it(`Operation login returns correct action`, () => {
    const apiMock = new MockAdapter(api);
    const dispatch = jest.fn();
    const checkAuth = Operation.login();

    apiMock
      .onPost(`/login`)
      .reply(200, {
        email: `korzan.va@mail.ru`,
        password: `fghjdk`,
      });

    return checkAuth(dispatch, () => {}, api)
      .then(() => {
        expect(dispatch).toHaveBeenCalledTimes(2);
        expect(dispatch.mock.calls[0][0]).toEqual({
          type: ActionType.REQUIRED_AUTHORIZATION,
          payload: `AUTH`});
        expect(dispatch.mock.calls[1][0]).toEqual({
          type: ActionType.AUTH_INFO,
          payload: {
            email: `korzan.va@mail.ru`,
            password: `fghjdk`,
          }});
      });
  });
});
