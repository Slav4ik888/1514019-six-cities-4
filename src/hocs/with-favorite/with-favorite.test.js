import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';

import withFavorite from './with-favorite.js';

import {AuthStatus} from '../../reducers/user/user.js';

import {NameSpace} from '../../reducers/name-space.js';
import {testOffer} from '../../mocks/test-offer.js';


const mockStore = configureStore([]);

const MockComp = () => <div/>;

const MockCompWrap = withFavorite(MockComp);


describe(`Snapshot <withFavorite /> correctly`, () => {
  it(`Render <withFavorite /> AuthStatus.AUTH`, () => {
    const store = mockStore({
      [NameSpace.USER]: {
        userStatus: AuthStatus.AUTH,
      }
    });
    const tree = renderer.create(
        <Provider store={store}>
          <MockCompWrap
            userStatus={AuthStatus.AUTH}
            offer={testOffer}
            onToggleFav={() => {}}
          />
        </Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Render <withFavorite /> AuthStatus.NO_AUTH`, () => {
    const store = mockStore({
      [NameSpace.USER]: {
        userStatus: AuthStatus.NO_AUTH,
      }
    });
    const tree = renderer.create(
        <Provider store={store}>
          <MockCompWrap
            userStatus={AuthStatus.NO_AUTH}
            offer={testOffer}
            onToggleFav={() => {}}
          />
        </Provider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});

// npm run test.jest -- -u with-favorite.test.js
// npm test with-favorite.test.js
