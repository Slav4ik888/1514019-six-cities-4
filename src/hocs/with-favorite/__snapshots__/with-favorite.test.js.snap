// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot <withFavorite /> correctly Render <withFavorite /> AuthStatus.AUTH 1`] = `<div />`;

exports[`Snapshot <withFavorite /> correctly Render <withFavorite /> AuthStatus.NO_AUTH 1`] = `<div />`;
